---
title: "MAGIC_imputation_analysis"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
getwd()
```

```{r pressure, echo=FALSE}
library(tidyr)
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(purrr)
library(devtools)
library(data.table)
library(Rmagic)
library(viridis)
library(phateR)
```

## Imputation Algorithm is not specified. The data upload had to be externally imputed with the desired imputation method. 

```{r}
# path_to_imputed_counts is the path to the csv file that contains the imputation results
read_data <- read.csv(file = "path_to_imputed_counts")
```

```{r}
# transpose the matrix back to genes X cells as SeuratObject expects to get
transpose_imputed_mat <- as.data.frame(t(as.matrix(read_data)))
```

```{r}
# make matrix gene x cells -> transpose
colnames(transpose_imputed_mat) <- transpose_imputed_mat[1,]
transpose_imputed_mat <- transpose_imputed_mat[-c(1), ]
transpose_imputed_mat
```

```{r}
# Initialize Seurat object with raw (un-normalized) data
seurat_obj_imputed <- CreateSeuratObject(counts = transpose_imputed_mat, project = "imputed_genes", min.cells = 3, min.features = 200)
seurat_obj_imputed
```

```{r}
# See what the dataset looks like 
seurat_obj_imputed[["RNA"]]@counts
```

We already see that more genes will be incorporated here.

## Standard pre processing workflow
We will use some common metrics to analyze our data:
 - The percentage of reads that map to the mitochondrial genome
 - number of unique genes in each cell

```{r}
# Here we will add the mitochondrial genes
seurat_obj_imputed[["percent.mt"]] <- PercentageFeatureSet(object = seurat_obj_imputed, pattern = "^MT")
```

View QC metrics using a violin plot use *VlnPlot*

```{r}
VlnPlot(seurat_obj_imputed, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

```{r}
head(seurat_obj_imputed@meta.data, 100)
seurat_obj_imputed@meta.data # 7865x4
```

Look at feature - feature relationship

```{r}
plot1 <- FeatureScatter(seurat_obj_imputed, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(seurat_obj_imputed, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot3 <- FeatureScatter(seurat_obj_imputed, feature1 = "percent.mt", feature2 = "nFeature_RNA")
# Display the plots
# To display side by side : plot1 + plot2 
plot1
plot2
plot3
```

We will now filter out data according to the results of the violin plot.

```{r}
# These numbers need be changed according to plot above
seurat_obj_imputed_filtered <- subset(seurat_obj_imputed, subset = nFeature_RNA > 12000 & nFeature_RNA < 16000 & percent.mt < 10 & nCount_RNA < 10000)
```

```{r}
seurat_obj_imputed_filtered@meta.data
```

## Normalize Data
Normalization with LogNormalize - multiply by 10000 and take log of result.

```{r}
seurat_obj_imputed_filtered <- NormalizeData(seurat_obj_imputed_filtered, normalization.method = "LogNormalize", scale.factor = 10000)
```

## Feature Selection 
We will now look for highly variable features. These features will be highly expressed in some cells and low in others. The default number of variable features is 2000. We are using the same number of features as we did for the non imputed data to make our comparison as viable as possible.

```{r}
seurat_obj_imputed_filtered <- FindVariableFeatures(seurat_obj_imputed_filtered, selection.method = "vst", nfeatures = 2000)
seurat_obj_imputed_filtered
```

Now we will plot the Variable Features and specifically label the top 10.

```{r}
# identify top 10 genes among the variable genes
top_10_imputed <- head(VariableFeatures(seurat_obj_imputed_filtered), 10)

# plot top 10 genes
plot1 <- VariableFeaturePlot(seurat_obj_imputed_filtered)
plot2 <- LabelPoints(plot = plot1, points = top_10_imputed, repel = TRUE)
plot1
plot2
```

## Scaling the data
We will scale the data before preforming any sort of dimensionality reduction.
This step will prevent the highly expressed genes from dominating in downstream analysis. 
Results will be stored in seurat_obj_imputed_filtered[["RNA"]]@scale.data.

```{r}
all.genes <- row.names(seurat_obj_imputed_filtered)
seurat_obj_imputed_filtered <- ScaleData(seurat_obj_imputed_filtered, features = all.genes)
```

## Preform linear dimensionality reduction
We will now preform PCA on the data.
The PCA will run based on the variable features we found earlier. 

```{r}
seurat_obj_imputed_filtered <- RunPCA(seurat_obj_imputed_filtered, features = VariableFeatures(seurat_obj_imputed_filtered))
```

```{r}
seurat_obj_imputed_filtered[["pca"]]
```

We will now examine and visualise the PCA results in a few ways.
We will use VisDimReduction(), DimPlot() and DimHeatmap() functions.
The positive and negative grouping refers to the genes in regards to the component.

```{r}
# Visualization of top genes associated with reduction components
VizDimLoadings(seurat_obj_imputed_filtered, dims = 1:2, reduction = "pca")
```

```{r}
# Plot PCA
DimPlot(seurat_obj_imputed_filtered, reduction = "pca")
```

```{r}
# Heatmap
DimHeatmap(seurat_obj_imputed_filtered, dims=1:15, cells=500, balanced = TRUE)
```

## Determine the 'dimensionality' of the dataset
Seurat clusters cells based on their PCA scores. 
The PC represents a "metafeature" that combines information across a correlated feature set.
We use here the JackStraw (JackStraw function) : 
Randomly permutes a subset of data, and calculates projected PCA scores for these 'random' genes. Then compares the PCA scores for the 'random' genes with the observed PCA scores to determine statistical significance. End result is a p-value for each gene's association with each principal component.

```{r}
seurat_obj_imputed_filtered <- JackStraw(seurat_obj_imputed_filtered, num.replicate = 100)
seurat_obj_imputed_filtered <- ScoreJackStraw(seurat_obj_imputed_filtered, dims = 1:20)
```

The JackStrawPlot() function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). 
‘Significant’ PCs will show a strong enrichment of features with low p-values (solid curve above the dashed line). 

```{r}
JackStrawPlot(seurat_obj_imputed_filtered, dims=1:15)
```

Now we'll use the ElbowPlot() - another tool for comparing the variation among the PC's.

```{r}
ElbowPlot(seurat_obj_imputed_filtered)
```

## Clustering Cells
1. Construct a KNN graph based on euclidean distance in PCA space (with edges drawn between cells with similar feature expression patterns). Input is the first 10 PC's (from conclusion above) - the defined dimensionality of the dataset.
2. Find Clusters - group cells together. 
The FindClusters() functions has a resolution (# of communities) argument that is usually set to 0.4-1.2 for 3K single-cell dataset - increases for larger datasets. 

```{r}
imputed_data_clusters <- FindNeighbors(seurat_obj_imputed_filtered,k.param = 10)
imputed_data_clusters <- FindClusters(imputed_data_clusters, resolution = seq(0.05,4,0.15))
```

We gave a sequence of resolutions to cluster by, and by that control the number 
of clusters.

```{r}
sapply(grep("res",colnames(imputed_data_clusters@meta.data),value = TRUE),
       function(x) length(unique(imputed_data_clusters@meta.data[,x])))
```

Different resolution values result in different amount of clusters.
Determine the resolution by taking into account the one chose for the original counts. 

```{r}
Idents(imputed_data_clusters) <- "RNA_snn_res.0.2"
head(Idents(imputed_data_clusters),5)
```

## Run non-linear dimensional reduction (UMAP)
We can use non linear dimensional reduction techniques to view and analyze our data.
The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space.
We will use the first 10 PC's (dims 1:10).
   
```{r}
imputed_UMAP <- RunUMAP(imputed_data_clusters, dims = 1:10)
DimPlot(imputed_UMAP, reduction = "umap", label = TRUE)
```

## Identification of all markers for each cluster

```{r}
DefaultAssay(imputed_UMAP) <- "RNA"
```

```{r}
imputed_markers <- FindAllMarkers(imputed_UMAP, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
```

```{r}
imputed_markers
```

```{r}
imputed_markers <- subset(imputed_markers, select = -c(gene))
```

```{r}
imputed_markers
View(imputed_markers)
```

In the output above:
 - avg_logFC : positive value indicates that the gene is highly expressed in the cluster
 - pct.1 : percentage of cells where the gene is detected in the cluster
 - pct.2 : percentage of cells where the gene is detected on average in the other clusters
 - p_val_adj : adjusted p-value for condition, based on bonferroni correction using all genes in the dataset, used to determine significance


## Adding gene annotations
Because we want to try and identify the clusters, it will be helpful to add to add a column with the gene annotation information. The data is in the annotation file.

```{r}
annotations <- read.csv("/Users/romigoldner/Desktop/Research Proposal/Masters/R Code/annotation.csv")
```

```{r}
annotations <- subset(annotations, select = c(gene_name, description))
View(annotations)
```

After we found the markers for each cluster, we will add the information from the annotations file per gene.
 
```{r}
imputed_markers <-imputed_markers %>% tibble::rownames_to_column(var = "gene")
cluster_markers <- left_join(imputed_markers, annotations, by = c("gene" = "gene_name"))
View(cluster_markers)
```

```{r}
# remove duplicate rows
cluster_markers <- cluster_markers[!duplicated(cluster_markers$gene), ]
```

## Evaluating marker genes
We would like to use these gene lists to see of we can identify which cell types these clusters identify with. 
Let’s take a look at the top genes for each of the clusters and see if that gives us any hints. 

```{r}
# Extract top 10 markers per cluster
top10_magic_cluster <- cluster_markers %>% group_by(cluster) %>% top_n(n=10, wt=avg_log2FC)

# Visualize top 10 markers per cluster
View(top10_magic_cluster)
```

The identification of the clusters will use the annotations attached and the Human Protein Atlas https://www.proteinatlas.org.

```{r}
# Rename all identities and adjust the cluster number according to results
seurat_imputed_UMAP <- RenameIdents(object = imputed_UMAP, 
    "0" = "XXXX",
    "1" = "XXXX",
    "2" = "XXXX",
    "3" = "XXXX",
    "4" = "XXXX",
    "5" = "XXXX",
    "6" = "XXXX",
    "7" = "XXXX",
    "8" = "XXXX",
    "9" = "XXXX",
    "10" = "XXXX",
    "11" = "XXXX",
    "12" = "XXXX",
    "13" = "XXXX",
    "14" = "XXXX")
```


```{r}
# Plot the UMAP wiht cluster identification
DimPlot(object = seurat_imputed_UMAP, 
        reduction = "umap", 
        label = TRUE,
        label.size = 3,
        repel = TRUE)
```
